test/e2e-bootstrap/bootstrap_test.go

```diff
@@ -0,0 +1,635 @@
+	mcfgv1.Install(scheme.Scheme)
+	apioperatorsv1alpha1.Install(scheme.Scheme)
+
+	baseTestManifests := loadBaseTestManifests(t)
+
+	cfg, err := testEnv.Start()
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, testEnv.Stop())
+	}()
+
+	clientSet := framework.NewClientSetFromConfig(cfg)
+
+	_, err = clientSet.Namespaces().Create(ctx, &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: openshiftConfigNamespace,
+		},
+	}, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	testCases := []struct {
+		name             string
+		manifests        [][]byte
+		waitForMasterMCs []string
+		waitForWorkerMCs []string
+	}{
+		{
+			name:             "With no additional manifests",
+			waitForMasterMCs: []string{"99-master-ssh", "99-master-generated-registries"},
+			waitForWorkerMCs: []string{"99-worker-ssh", "99-worker-generated-registries"},
+		},
+		{
+			name: "With a featuregate manifest",
+			manifests: [][]byte{
+				[]byte(`apiVersion: config.openshift.io/v1
+kind: FeatureGate
+metadata:
+  name: cluster
+spec:
+  featureSet: TechPreviewNoUpgrade`),
+			},
+			waitForMasterMCs: []string{"99-master-ssh", "99-master-generated-registries", "98-master-generated-kubelet"},
+			waitForWorkerMCs: []string{"99-worker-ssh", "99-worker-generated-registries", "98-worker-generated-kubelet"},
+		},
+		{
+			name: "With a featuregate manifest and master kubelet config manifest",
+			manifests: [][]byte{
+				[]byte(`apiVersion: config.openshift.io/v1
+kind: FeatureGate
+metadata:
+  name: cluster
+spec:
+  featureSet: TechPreviewNoUpgrade`),
+				[]byte(`apiVersion: machineconfiguration.openshift.io/v1
+kind: KubeletConfig
+metadata:
+  name: master-kubelet-config
+spec:
+  machineConfigPoolSelector:
+    matchLabels:
+      pools.operator.machineconfiguration.openshift.io/master: ""
+  kubeletConfig:
+    podsPerCore: 10
+    maxPods: 250
+    systemReserved:
+      cpu: 1000m
+      memory: 500Mi
+    kubeReserved:
+      cpu: 1000m
+      memory: 500Mi
+`),
+			},
+			waitForMasterMCs: []string{"99-master-ssh", "99-master-generated-registries", "99-master-generated-kubelet"},
+			waitForWorkerMCs: []string{"99-worker-ssh", "99-worker-generated-registries"},
+		},
+		{
+			name: "With a worker kubelet config manifest",
+			manifests: [][]byte{
+				[]byte(`apiVersion: machineconfiguration.openshift.io/v1
+kind: KubeletConfig
+metadata:
+  name: worker-kubelet-config
+spec:
+  machineConfigPoolSelector:
+    matchLabels:
+      pools.operator.machineconfiguration.openshift.io/worker: ""
+  kubeletConfig:
+    podsPerCore: 10
+    maxPods: 250
+    systemReserved:
+      cpu: 1000m
+      memory: 500Mi
+    kubeReserved:
+      cpu: 1000m
+      memory: 500Mi
+`),
+			},
+			waitForMasterMCs: []string{"99-master-ssh", "99-master-generated-registries"},
+			waitForWorkerMCs: []string{"99-worker-ssh", "99-worker-generated-registries", "99-worker-generated-kubelet"},
+		},
+	}
+
+	for _, tc := range testCases {
```

Comment:
It's a bit unclear to me, are these individual test cases(applying manifests) running during bootstrap and we are later on comparing rendered config of bootstrap version and rendered config available during real cluster. How we are ensuring that each manifests are getting applied with fresh bootstrap process?

Reply:
This might be better explained via a call, does your team have an office hours I could attend to demo this test and explain properly how it works?

The `envtest` tool I've set up in this PR runs an etcd and kubeapiserver locally on your machine (or in CI), so there isn't actually a real cluster per se.

https://github.com/openshift/machine-config-operator/blob/f4aad781788b1a864e7bfe6cbd0572bffa590562/test/e2e-bootstrap/bootstrap_test.go#L82-L86

This test suite runs the controllers from the MCC against this local kubeapiserver. It then creates all the resources that would normally be created during bootstrap, which the MCC controllers react to, and render the config as they would do in a real cluster.

https://github.com/openshift/machine-config-operator/blob/f4aad781788b1a864e7bfe6cbd0572bffa590562/test/e2e-bootstrap/bootstrap_test.go#L262-L280

Alongside this, we take the same resources and put them into a temp folder, which we then run the bootstrap against.

https://github.com/openshift/machine-config-operator/blob/f4aad781788b1a864e7bfe6cbd0572bffa590562/test/e2e-bootstrap/bootstrap_test.go#L198-L221

Because this isn't a real cluster, and the only controllers running are the MCC controllers, we then shut down the controllers between each test and restart them. And we clean the environment by deleting all of the resources that we created.

https://github.com/openshift/machine-config-operator/blob/f4aad781788b1a864e7bfe6cbd0572bffa590562/test/e2e-bootstrap/bootstrap_test.go#L186-L187

One of the really cool things about this style of testing is you can see how your controller reacts to real API resources, without fakes, without a real cluster :)

Reply:
Thanks for explaining. We don't have office hours scheduled so far, we can plan a meeting but that can earliest happen next week.